src/components/include/cbuf.h

```diff
@@ -405,241 +341,131 @@ __cbuf_alloc_meta_inconsistent(struct cbuf_alloc_desc *d, struct cbuf_meta *m)
 
-	CBUF_TAKE();
-	cm = cbuf_vect_lookup_addr(cbid_to_meta_idx(cbid), tmem);
+	cm = cbuf_vect_lookup_addr(cbid_to_meta_idx(id));
 	/* 
 	 * If this assertion triggers, one possibility is that you did
 	 * not successfully map it in (cbufp2buf or cbufp_alloc).
 	 */
-	assert(cm->nfo.c.refcnt);
-	owner = cm->nfo.c.flags & CBUFM_OWNER;
-
+	assert(CBUFM_GET_REFCNT(cm));
+	owner = CBUF_OWNER(cm);
+	tmem = CBUF_TMEM(cm);
+	inconsistent = CBUF_INCONSISENT(cm);
+	assert(!inconsistent);
 	if (tmem) {
-		assert(owner); /* Shouldn't be calling free... */
-		struct cbuf_alloc_desc *d = __cbuf_alloc_lookup(cm->nfo.c.ptr);
-		struct cbuf_alloc_desc *fl;
-
-		assert(d && !__cbuf_alloc_meta_inconsistent(d, cm));
-
-		fl = d->flhead;
-		assert(fl);
-		ADD_LIST(fl, d, next, prev);
-		cm->owner_nfo.thdid = 0;
+		assert(owner);
+		fl = __cbuf_freelist_get(cm->sz << PAGE_ORDER);
+		CBUFM_SET_NEXT(cm, CBUFM_GET_NEXT(fl));
+		cos_mem_fence();
+		CBUFM_DEC_REFCNT(cm);
+		cos_mem_fence();
+		target = (unsigned long long *)(&fl->next_flag);
+		old    = (unsigned long long *)(&head.next_flag);
+		update = (unsigned long long *)(&new_head.next_flag);
+		new_head.next_flag = cm;
+		do {
+			*old = *target;
+			new_head.cbid.tag  = fl->cbid.tag+1;
+		} while(unlikely(!cos_dcas(target, *old, *update)));
 	}
-	/* do this last, so that we can guarantee the manager will not steal the cbuf before now... */
-	cm->nfo.c.refcnt--;
-	if (tmem) cos_comp_info.cos_tmem_available[COMP_INFO_TMEM_CBUF]++;
-	else      relinq = cm->nfo.c.flags & CBUFM_RELINQ;
-	CBUF_RELEASE();
-	
+	else CBUFM_DEC_REFCNT(cm);      
+	relinq = CBUF_RELINQ(cm);
 	/* Does the manager want the memory back? */
-	if (tmem) {
-		if (unlikely(cos_comp_info.cos_tmem_relinquish[COMP_INFO_TMEM_CBUF])) {
-			cbuf_c_delete(cos_spd_id(), cbid);
-			assert(lock_contested(&cbuf_lock) != cos_get_thd_id());
-			return;
-		}
-	} else if (unlikely(relinq)) {
-		cbufp_delete(cos_spd_id(), cbid);
-		assert(lock_contested(&cbuf_lock) != cos_get_thd_id());
-		return;
-	}
-	
+	if (unlikely(relinq)) cbufp_delete(cos_spd_id(), id);	
 	return;
 }
 
-static inline void
-cbufp_deref(cbufp_t cbid) 
-{ 
-	u32_t id;
-	int tmem;
-	
-	cbuf_unpack(cbid, &id, &tmem);
-	assert(tmem == 0);
-	__cbuf_done((int)id, 0);
-}
-
-static inline void
-cbuf_free(cbuf_t cb)
-{
-	u32_t id;
-	int tmem;
-	cbuf_unpack(cb, &id, &tmem);
-	assert(tmem == 1);
-	__cbuf_done((int)id, 1);
-}
-
-/* 
- * Is it a cbuf?  If so, what's its id? 
- *
- * This only works in the allocating component (with the freelist
- * descriptor).
- */
-static inline int 
-cbuf_id(void *buf)
-{
-	u32_t idx = ((u32_t)buf) >> PAGE_ORDER;
-	struct cbuf_alloc_desc *d;
-	int id;
-
-	CBUF_TAKE();
-	d  = __cbuf_alloc_lookup(idx);
-	id = (likely(d)) ? d->cbid : 0;
-	CBUF_RELEASE();
-	assert(lock_contested(&cbuf_lock) != cos_get_thd_id());
-
-	return id;
-}
-
 /* Is the pointer a cbuf?  */
 static inline int 
 cbuf_is_cbuf(void *buf)
 {
-	return cbuf_id(buf) != 0;
+	return false;   //???
```

Comment:
What is `false`?  I don't think that should be defined anywhere.  `return 0;`.

This is used to see if a passed in memory region is actually backed by a cbuf.  Once Gedare's work is done, this will always be true, so we can probably remove this function.